crates/lunatic-sqlite-api/src/lib.rs

```diff
@@ -0,0 +1,264 @@
+        ///     unsafe {
+        ///         // the leftmost half contains the length
+        ///         let length = (ptr >> 32) as usize;
+        ///         // the rightmost half contains the pointer
+        ///         let ptr = 0x00000000FFFFFFFF & ptr;
+        ///         Vec::from_raw_parts(ptr as *mut u8, length, length)
+        ///     }
+        /// }
+        /// ```
+        ///
+        ///
+
+        /// looks up the value of the last error, encodes an `SqliteError` via bincode
+        /// and writes it into a guest allocated Vec<u8>
+        /// Returns a composite length + pointer to the data (see explanation above)
+        pub fn last_error(connection_id: u64) -> u64;
+
+        /// reads the column under index `col_idx` encodes a `SqliteValue` via bincode
+        /// and writes it into a guest allocated Vec<u8>
+        /// Returns a composite length + pointer to the data (see explanation above)
+        pub fn read_column(statement_id: u64, col_idx: u32) -> u64;
+
+        /// reads the next row, encodes a `SqliteRow` via bincode
+        /// and writes it into a guest allocated Vec<u8>
+        pub fn read_row(statement_id: u64) -> u64;
+
+        /// looks up the name of the column under index `col_idx`, encodes a `String` via bincode
+        /// and writes it into a guest allocated Vec<u8>
+        pub fn column_name(statement_id: u64, col_idx: u32) -> u64;
+
+        /// looks up the value of the last error, encodes a `Vec<String>` via bincode
+        /// and writes it into a guest allocated Vec<u8>
+        pub fn column_names(statement_id: u64) -> u64;
+
+        /// sets a custom guest allocator for a connection identified by connection_id
+        /// this should not be used unless using the library outside out lunatic-sql which
+        /// would mean that the allocator function is missing and you need to provide your own
+        pub fn set_custom_guest_allocator(
+            connection_id: u64,
+            allocator_name_ptr: u32,
+            allocator_name_len: u32,
+        );
+    }
+}
+
+pub const SQLITE_OK: u32 = 0;
+pub const SQLITE_ERROR: u32 = 1;
+pub const SQLITE_INTERNAL: u32 = 2;
+pub const SQLITE_PERM: u32 = 3;
+pub const SQLITE_ABORT: u32 = 4;
+pub const SQLITE_BUSY: u32 = 5;
+pub const SQLITE_LOCKED: u32 = 6;
+pub const SQLITE_NOMEM: u32 = 7;
+pub const SQLITE_READONLY: u32 = 8;
+pub const SQLITE_INTERRUPT: u32 = 9;
+pub const SQLITE_IOERR: u32 = 10;
+pub const SQLITE_CORRUPT: u32 = 11;
+pub const SQLITE_NOTFOUND: u32 = 12;
+pub const SQLITE_FULL: u32 = 13;
+pub const SQLITE_CANTOPEN: u32 = 14;
+pub const SQLITE_PROTOCOL: u32 = 15;
+pub const SQLITE_EMPTY: u32 = 16;
+pub const SQLITE_SCHEMA: u32 = 17;
+pub const SQLITE_TOOBIG: u32 = 18;
+pub const SQLITE_CONSTRAINT: u32 = 19;
+pub const SQLITE_MISMATCH: u32 = 20;
+pub const SQLITE_MISUSE: u32 = 21;
+pub const SQLITE_NOLFS: u32 = 22;
+pub const SQLITE_AUTH: u32 = 23;
+pub const SQLITE_FORMAT: u32 = 24;
+pub const SQLITE_RANGE: u32 = 25;
```

Comment:
We could move this to `lunatic-sql` as part of the sqlite diesel bindings because that's ultimately who's using these constants. In the lunatic-sqlite-api crate we actually only need 2 constant - SQLITE_DONE and SQLITE_ROW. The rest is used by the ORM. Since breaking API changes would have to be handled properly in the ORM bindings anyway I think we can safely move them there. Especially if we want to create a more database-agnostic interface